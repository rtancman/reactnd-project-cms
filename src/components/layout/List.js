import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Link } from "react-router-dom"
import moment from 'moment'
import sortBy from 'sort-by'
import CommentIcon from '@material-ui/icons/Comment';
import EventIcon from '@material-ui/icons/Event';
import './List.css'
import Vote from '../vote'

const orderByOptions = [
  {label: 'Popular', value: '-voteScore'},
  {label: 'Category', value: 'category'},
  {label: 'Recent', value: '-timestamp'},
]

class BaseList extends Component {
  static propTypes = {
    items: PropTypes.array.isRequired,
    title: PropTypes.string,
  }

  state = {
    orderBy: false,
  }

  orderBy(option) {
    this.setState({
      orderBy: option
    })
  }
}

class ListContent extends BaseList {

  state = {
    orderBy: '-voteScore'
  }

  render() {
    const { items, title } = this.props
    const { orderBy } = this.state
    let showingPosts = items

    showingPosts.sort(sortBy(orderBy))

    return (
      <div className="list_content">
        <h2>{ title }</h2>
        <div className="row list_content__filter">
          { orderByOptions.map((option) => (
            <div data-filter-option-value={ option.value } onClick={ () => this.orderBy(option.value)} key={ option.value } className={ `col-xs list_content__filter__item ${option.value === orderBy? 'list_content__filter__item--selected' : ''}` }>
              <span>{ option.label }</span>
            </div>
          )) }
        </div>
        {showingPosts.map((item) => (
          <div className="list_content__item" key={item.id}>
            <div className="list_content__item__title">
              <Link className="link" to={`/${item.category}/${item.id}`}>{item.title}</Link>
              <div className="list_content__item__title--vote">
                <Vote
                  type='post' 
                  id={item.id}
                  total={item.voteScore || 0}
                />
              </div>
            </div>
            <div className="list_content__item__info">
              <p>
                <EventIcon style={{ fontSize: 10 }} title='Create at' alt='Create at' /> { moment(item.timestamp).format('MM/DD/YYYY') } <CommentIcon style={{ fontSize: 10 }} title='Total comments' alt='Total comments' /> { item.commentCount } - in <Link className="link" to={`/${item.category}`}>{ item.category }</Link> - by <span className="list_posts__item__info--author">{ item.author }</span>
              </p>
            </div>
          </div>
        ))}
      </div>
    )
  }
}

const orderByCategoryOptions = [
  {label: 'Default', value: false},
  {label: 'Name', value: 'path'},
]
class List extends BaseList {
  render() {
    const { items, title } = this.props
    const { orderBy } = this.state
    let showingItems = items
    if ( orderBy ){
      showingItems.sort(sortBy(orderBy))
    }

    return (
      <div className="list_content">
        <h2>{ title }</h2>
        <div className="row list_content__filter">
          { orderByCategoryOptions.map((option) => (
            <div data-filter-option-value={ option.value } onClick={ () => this.orderBy(option.value)} key={ option.value } className={ `col-xs list_content__filter__item ${option.value === orderBy? 'list_content__filter__item--selected' : ''}` }>
              <span>{ option.label }</span>
            </div>
          )) }
        </div>
        {showingItems.map((item) => (
          <div className="list_content__item list_content__item--category" key={item.path}>
            <div className="list_content__item__title">
              <Link className="link" to={`/${item.path}`}>{item.name}</Link>
            </div>
          </div>
        ))}
      </div>
    )
  }
}

List.propTypes = {
  items: PropTypes.array.isRequired,
}

export { ListContent, List }