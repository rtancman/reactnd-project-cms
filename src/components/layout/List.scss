@import '../_mediaqueries.scss';

.list_content {
  &__item {
    margin-top: 30px;
    margin-bottom: 10px;
    padding-bottom: 20px;
    border-bottom: 0.5px solid #E1E1E1;
    
    &:last-child {
      border-bottom: 0;
    }

    &--category {
      padding-bottom: 37px;
    }

    &__title {
      font-size: 24px;
      letter-spacing: -1.25px;
      line-height: 28px;
      font-weight: 600;
      display: flex;

      @include small {
        flex-direction: column;
      }

      a {
        flex: 1 1;
        @include small {
          margin-bottom: 10px;
        }
      }
      &--vote {
        justify-content: flex-end;
        font-size: 17px;
        .vote__bar__total {
          padding: 0 10px;
        }
        @include small {
          flex: 1 1;
          margin-bottom: 10px;
        }
      }
    }
  
    &__info {
      color: #999;
      font-size: 11px;
      grid-column: 2;
      line-height: 11px;
      margin: 6px 0 0;
      text-transform: uppercase;
      p {
        margin: 0;
      }
    }
  }

  &__filter {
    margin: 0;

    &__item {
      padding: 12px 0 9px 0;
      border: 0.5px solid #e0e0e0;
      font-weight: bold;
      font-size: 12px;
      color: #666;
      text-transform: uppercase;
      text-align: center;
      cursor: pointer;
      line-height: 1;

      &:first-child {
        border-radius: 3px 0 0 3px;
      }

      &:last-child {
        border-radius: 0 3px 3px 0;
      }

      &--selected {
        border-color: #000;
        color: #000;
      }
    }
  }
}