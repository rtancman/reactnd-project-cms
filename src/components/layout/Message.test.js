import React from 'react'
import { shallow, mount } from 'enzyme'
import { MemoryRouter } from 'react-router'
import { ShowMessage } from 'components/layout/Message'

describe('ShowMessage Component', () => {
  describe('when component renders', () => {
    it('without crashing', () => {
      const props = {
        message: 'Message',
        variant: 'success',
        open: true,
      }
      const wrapper = shallow(<ShowMessage {...props} />)

      expect(wrapper).toMatchSnapshot()
    })

    it('with success message', () => {
      const props = {
        message: 'Message success',
        variant: 'success',
        open: true,
      }
      const wrapper = mount(<ShowMessage {...props} />)
      
      const snackBar = wrapper.find('Snackbar')
      expect(snackBar.length).toBe(1)
      const snackBarMessage = snackBar.find('span.client__snackbar')
      expect(snackBarMessage.length).toBe(1)
      expect(snackBarMessage.text()).toBe('Message success')
      expect(snackBarMessage.find('CheckCircleIcon').length).toBe(1)
      const snackBarAction = snackBar.find('IconButton')
      expect(snackBarAction.length).toBe(1)
      expect(snackBarAction.find('CloseIcon').length).toBe(1)
      snackBarAction.simulate('click')
      expect(snackBar.find('CloseIcon').length).toBe(1)
      expect(wrapper.state('open')).toEqual(false)
    })

    it('with warning message', () => {
      const props = {
        message: 'Message warning',
        variant: 'warning',
        open: true,
      }
      const wrapper = mount(<ShowMessage {...props} />)
      
      const snackBar = wrapper.find('Snackbar')
      expect(snackBar.length).toBe(1)
      const snackBarMessage = snackBar.find('span.client__snackbar')
      expect(snackBarMessage.length).toBe(1)
      expect(snackBarMessage.text()).toBe('Message warning')
      expect(snackBarMessage.find('WarningIcon').length).toBe(1)
      const snackBarAction = snackBar.find('IconButton')
      expect(snackBarAction.length).toBe(1)
      expect(snackBarAction.find('CloseIcon').length).toBe(1)
      snackBarAction.simulate('click')
      expect(snackBar.find('CloseIcon').length).toBe(1)
      expect(wrapper.state('open')).toEqual(false)
    })

    it('with error message', () => {
      const props = {
        message: 'Message error',
        variant: 'error',
        open: true,
      }
      const wrapper = mount(<ShowMessage {...props} />)
      
      const snackBar = wrapper.find('Snackbar')
      expect(snackBar.length).toBe(1)
      const snackBarMessage = snackBar.find('span.client__snackbar')
      expect(snackBarMessage.length).toBe(1)
      expect(snackBarMessage.text()).toBe('Message error')
      expect(snackBarMessage.find('ErrorIcon').length).toBe(1)
      const snackBarAction = snackBar.find('IconButton')
      expect(snackBarAction.length).toBe(1)
      expect(snackBarAction.find('CloseIcon').length).toBe(1)
      snackBarAction.simulate('click')
      expect(snackBar.find('CloseIcon').length).toBe(1)
      expect(wrapper.state('open')).toEqual(false)
    })

    it('with info message', () => {
      const props = {
        message: 'Message info',
        variant: 'info',
        open: true,
      }
      const wrapper = mount(<ShowMessage {...props} />)
      
      const snackBar = wrapper.find('Snackbar')
      expect(snackBar.length).toBe(1)
      const snackBarMessage = snackBar.find('span.client__snackbar')
      expect(snackBarMessage.length).toBe(1)
      expect(snackBarMessage.text()).toBe('Message info')
      expect(snackBarMessage.find('InfoIcon').length).toBe(1)
      const snackBarAction = snackBar.find('IconButton')
      expect(snackBarAction.length).toBe(1)
      expect(snackBarAction.find('CloseIcon').length).toBe(1)
      snackBarAction.simulate('click')
      expect(snackBar.find('CloseIcon').length).toBe(1)
      expect(wrapper.state('open')).toEqual(false)
    })
  })
})