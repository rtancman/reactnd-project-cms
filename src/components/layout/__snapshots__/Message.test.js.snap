// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowMessage Component when component renders without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ShowMessage
    message="Message"
    open={true}
    variant="success"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "anchorOrigin": Object {
        "horizontal": "center",
        "vertical": "top",
      },
      "autoHideDuration": 3000,
      "children": <WithStyles(MySnackbarContent)
        message="Message"
        onClose={[Function]}
        variant="success"
/>,
      "onClose": [Function],
      "open": true,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "message": "Message",
        "onClose": [Function],
        "variant": "success",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "anchorOrigin": Object {
          "horizontal": "center",
          "vertical": "top",
        },
        "autoHideDuration": 3000,
        "children": <WithStyles(MySnackbarContent)
          message="Message"
          onClose={[Function]}
          variant="success"
/>,
        "onClose": [Function],
        "open": true,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "message": "Message",
          "onClose": [Function],
          "variant": "success",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
